java/jdbc/statement-interceptor/interceptor/Readme.md

```diff
@@ -0,0 +1,110 @@
+# SQL Interceptor
+![oveview](assets/img/overview.png)
+
+This project demonstrates how the _TraceEventListener_ feature can be used for security
```

Comment:
This project demonstrates how the new 23ai Oracle JDBC feature [TraceEventListener|https://docs.oracle.com/en/database/oracle/oracle-database/23/jajdb/oracle/jdbc/TraceEventListener.html] can be used for security enhancement of Java applications. The [TraceEventListener|https://docs.oracle.com/en/database/oracle/oracle-database/23/jajdb/oracle/jdbc/TraceEventListener.html] is a callback that can be registered for every roundtrip made to the Database. This particular implementation can intercept any SQL statement issued by an Oracle JDBC client application and, based on given rules, allow the statement to proceed and reach the remote server.

This TraceEventListenerProvider registers a listener that analyzes all round-trips made to the remote server. All statements are then analysed by a set of "statement rules" that "authorize" or not the request to go on.

A statement rule (StatementRule) is a class that processes a SQL statement as input and executes an action if that statement matches a rule. Rules can be used to identify security risks, such as a SQL injection attacks.

The advantage of such interceptor is that it can be plugged-in transparently to any existing application without any code change. It does not require any extra layer of software (like  etting up a firewall or proxy), it may help to prevent some DOS by just forbidding statements to reach the Oracle Database server.